src/ucp/core/ucp_request.inl

```diff
@@ -15,58 +15,77 @@
+    (((_flags) & UCP_REQUEST_FLAG_BLOCKING)        ? 'b' : '-'), \
+    (((_flags) & UCP_REQUEST_FLAG_EXPECTED)        ? 'e' : '-'), \
+    (((_flags) & UCP_REQUEST_FLAG_LOCAL_COMPLETED) ? 'L' : '-'), \
+    (((_flags) & UCP_REQUEST_FLAG_CALLBACK)        ? 'c' : '-'), \
+    (((_flags) & UCP_REQUEST_FLAG_RECV)            ? 'r' : '-'), \
+    (((_flags) & UCP_REQUEST_FLAG_SYNC)            ? 's' : '-'), \
+    (((_flags) & UCP_REQUEST_FLAG_RNDV)            ? 'v' : '-')
+
 
-    if (req != NULL) {
-        VALGRIND_MAKE_MEM_DEFINED(req + 1,  worker->context->config.request.size);
+/* defined as a macro to print the call site */
+#define ucp_request_get(_worker) \
+    ({ \
+        ucp_request_t *req = ucs_mpool_get_inline(&(_worker)->req_mp); \
+        if (req != NULL) { \
+            VALGRIND_MAKE_MEM_DEFINED(req + 1, \
+                                      (_worker)->context->config.request.size); \
+            ucs_trace_req("allocated request %p", req); \
+        } \
+        req; \
+    })
+
+#define ucp_request_complete(_req, _cb, _status, ...) \
+    { \
+        (_req)->status = (_status); \
+        if (ucs_likely((_req)->flags & UCP_REQUEST_FLAG_CALLBACK)) { \
+            (_req)->_cb((_req) + 1, (_status), ## __VA_ARGS__); \
+        } \
+        if (ucs_unlikely(((_req)->flags  |= UCP_REQUEST_FLAG_COMPLETED) & \
+                         UCP_REQUEST_FLAG_RELEASED)) { \
+            ucp_request_put(_req); \
+        } \
     }
-    return req;
-}
 
 static UCS_F_ALWAYS_INLINE void
-ucp_request_put(ucp_request_t *req, ucs_status_t status)
+ucp_request_put(ucp_request_t *req)
 {
-    req->status = status;
-    if ((req->flags |= UCP_REQUEST_FLAG_COMPLETED) & UCP_REQUEST_FLAG_RELEASED) {
-        /* Release should not be called for external requests */
-        ucs_assert(!(req->flags & UCP_REQUEST_FLAG_EXTERNAL));
-        ucs_mpool_put(req);
-    }
+    ucs_trace_req("put request %p", req);
+    ucs_assert(req->flags & UCP_REQUEST_FLAG_COMPLETED);
+    ucs_mpool_put_inline(req);
 }
 
 static UCS_F_ALWAYS_INLINE void
 ucp_request_complete_send(ucp_request_t *req, ucs_status_t status)
 {
-    ucs_trace_data("completing send request %p (%p), %s", req, req + 1,
-                   ucs_status_string(status));
-    req->send.cb(req + 1, status);
-
+    ucs_trace_req("completing send request %p (%p) "UCP_REQUEST_FLAGS_FMT" %s",
+                  req, req + 1, UCP_REQUEST_FLAGS_ARG(req->flags),
+                  ucs_status_string(status));
     UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_UCP_TX,
                           "ucp_request_complete_send",
                           req, 0);
-    ucp_request_put(req, status);
+    ucp_request_complete(req, send.cb, status);
 }
 
 static UCS_F_ALWAYS_INLINE void
-ucp_request_complete_recv(ucp_request_t *req, ucs_status_t status,
-                          ucp_tag_recv_info_t *info)
+ucp_request_complete_recv(ucp_request_t *req, ucs_status_t status)
```

Comment:
It might be better names below to avoid confusion that something sending or receiving.
ucp_request_complete_recv -> ucp_recv_request_complete
ucp_request_complete_send -> ucp_send_request_complete